src/docs/asciidoc/distributed_sql.adoc

```diff
@@ -0,0 +1,211 @@
+optimizes it using Apache Calcite. The result is an executable plan that
+is cached and reused by the subsequent executions of the same statement.
+
+The plan contains a tree of query fragments. A query fragment is a tree of
+operators that could be executed on a single member independently. Child
+fragments supply data to parent fragments, possibly through a network, until
+the root fragment is reached. The root fragment returns query results to a
+user.
+
+Only `IMap` objects are supported at the moment.
+
+=== Querying IMap
+
+==== Names
+
+The engine exposes `IMap` objects as tables in the predefined `partitioned`
+schema using exact names. This schema is in the engine search path so that
+`IMap` objects could be accessed with or without the schema name.
+
+Schema and table names are case sensitive. For example, the `IMap` with the name
+`employee` can be accessed as `employee` or `partitioned.employee`, but not as
+`Employee`.
+
+==== Fields
+
+The engine reads the first local entry of the `IMap` to get the list of fields
+accessible from the SQL. The engine extracts field metadata (name, type)
+from the key and the value, and combines them into a final list of fields.
+
+Field metadata extraction logic depends on the serialization format, as described
+below.
+
+-  For Java objects (`Serializable`, `DataSerializable`, `IdentifiedDataSerializable`),
+the object is deserialized if needed and then analyzed using reflection. Only public
+fields and getters are taken in count. A public field is exposed using its exact name.
+A getter is exposed using the JavaBean naming conventions. E.g. `getName` is exposed as
+`name` and `isActive` is exposed as `active`
+- For `Portable` objects, only the fields that are written in the `Portable.writePortable`
+method are exposed using their exact names
+- `HazelcastJsonValue` objects cannot be queried at the moment. This limitation will be
+removed in future releases
+
+Field names are case-sensitive.
+
+If the `IMap` does not have local entries on the member where the query is started,
+then the list of fields cannot be resolved, and an exception is thrown.
+
+We refer to the key and the value as *top-level* objects. The key is exposed as
+`__key` field, and the value is exposed as `this` field.
+
+If a top-level object does not have nested fields, then it is exposed as a normal
+field. Otherwise, it is exposed a *hidden field*. Hidden fields behave similarly
+to normal fields, but are not returned by `SELECT * ...` queries.
+
+If both the key and the value have fields with the same name, then the field of the
+value will be used.
+
+==== Indexes
+
+The engine could use `IMap` indexes to speed up the execution of certain queries.
+
+`SORTED` and `HASH` indexes are supported.
+
+==== HD
+
+`IMap` objects with the `InMemoryFormat.NATIVE` could only be queried if they have
+at least one index, and the property `hazelcast.hd.global.index.enabled` is set to
+`true`.
+
+=== Supported Features
```

Comment:
Could we please arrange it in a table? I want to have a very clear vision on what is supported and what not. The table should have 3 columns - "Feature" (or no caption), "Supported", "Notes".
In the table, you would create the sections by joining the column in a the row making as a headline for the sections like you have here "Data types", "Operators", "Functions" ...
And for the "Supported" column, please use a green check icon and red cross icon. I want to see a clear visual without reading, just red/green to punch me in the eye.

Reply:
I arranged `Data Types` in the table. It is Java-specific in some sense because it demonstrates mapping between SQL and Java types. There is only Java because other platforms are not supported currently.

Regarding operators/functions - this is not how SQL documentation is typically organized for many reasons:
- There is no fixed set of features that a vendor should have. No vendor is ANSI compliant
- There are hundreds of different functions, operators and keywords that are vendor-specific

Therefore, no-one organizes their docs in a form "we support this, but do not support this". Instead, every keyword and every function have a dedicated page, where the syntax and semantics are explained. And there is an index with the full list of supported keywords and functions.

The approach with checks and crosses could work well for, say, JDBC, where there is a big list of features that a vendor *must* support to be compliant with different JDBC compliance levels. 

I rearranged the section. Data types are now separated. Operators and functions are merged into a single block since we do not have a rich feature set, and detailed documentation of every keyword/function is currently out of scope.

WDYT?